src/server.c

```diff
@@ -193,6 +193,38 @@ void srv_set_dyncookie(struct server *s)
 	HA_RWLOCK_RDUNLOCK(PROXY_LOCK, &p->lock);
 }
 
+static void srv_set_addr_desc(struct server *s)
+{
+	struct proxy *p = s->proxy;
+	char *key;
+
+	key = sa2str(&s->addr, s->svc_port, s->flags & SRV_F_MAPPORTS);
+
+	if (strcmp(key, s->addr_desc) == 0) {
+		free(key);
+		return;
+	}
+
+	HA_RWLOCK_WRLOCK(PROXY_LOCK, &p->lock);
+	ebpt_delete(&s->addr_node);
+	HA_RWLOCK_WRUNLOCK(PROXY_LOCK, &p->lock);
+
+	free(s->addr_desc);
+	s->addr_desc = key;
+	s->addr_node.key = key;
+
+	// TODO: should this use a dedicated lock?
+	HA_RWLOCK_WRLOCK(PROXY_LOCK, &p->lock);
```

Comment:
I'm not entirely sure about the locking here. Would it be better to hold the lock over the entire operation, or just while we are actually manipulating the tree? And should this use the existing proxy lock, or should there be a separate dedicated lock for `used_server_addr`?